# CM4125 Topic 6 Lab 1, Joins and Aggregations (Guided)

The `rating.csv` dataset contains user ratings of anime

The data set is structed as follows:
    
| user_id | anime_id | rating |
| ------: | -------: | -----: |
|       1 |    11757 |     10 |
|       2 |    11771 |     10 |
|       3 |       20 |      8 |
|       3 |      154 |      6 |
|       3 |      170 |      9 |
|     ... |      ... |    ... |

Notice that no column is a unique index

In fact, `user_id` indicates the user doing the rating, `anime_id` indicates the anime, and `rating` is the user's rating of that anime from `1` to `10`

Import this dataset, removing in advance all rows with missing data (you should have $5484219$ rows of data)

In [1]:
## Load the data and then do something similar to the .notnull operation in Python
## This will only load the entries that have user_id and anime_id and rating!

user_id  anime_id  rating
0              1     11757      10
1              2     11771      10
2              3        20       8
3              3       154       6
4              3       170       9
...          ...       ...     ...
5484214    73515     14345       7
5484215    73515     16512       7
5484216    73515     17187       9
5484217    73515     22145      10
5484218    73516       790       9

[5484219 rows x 3 columns]

Let's reorder the dataset as follows:

1. Group the data by `user_id`
2. Aggregate using the `count` operation
3. Sort by count of rating (descending)

In [2]:
## To do this you need to first group all your data by user_id, 
## Then you need to consider that you want the aggregation by count of ratings
## Finally, you sort the table in descending order for the rating counts

count
user_id       
42635     3243
51693     2507
57620     2422
59643     2314
7345      2192
...        ...
46799        1
46795        1
46794        1
46765        1
73516        1

[69180 rows x 1 columns]

If you did all of these steps correctly, now you should have a dataset that tells you how many animes each of the $69180$ users has rated

Moreover, your `user_id` column should be now the index!

**Q. Which user (by ID) has rated the most anime?**

Answer:

Now you will do the following:

1. Create a new dataframe called `anime_ratings` which shows the number of times each anime was rated and it's mean rating
2. Rename column `count` as `Number of Ratings` and `mean` to `Average Rating`
3. Drop any anime which received fewer than $100$ ratings and sort by `Number of Ratings` (descending)
4. Round the average ratings to 2 decimal places

In [3]:
## You need to create a new grouping, but now based on anime_id
## Moreover, you need to see if it is possible to aggregate, both with count and mean!
## Finally, filter out rows with less than 100 ratings in the count, which tou renamed as "Number of ratings"

Average Rating  Number of Ratings
anime_id                                   
1535                8.83              34226
11757               8.14              26310
16498               8.73              25290
1575                8.93              24126
6547                8.55              23565
...                  ...                ...
2249                5.76                100
18771               6.97                100
360                 6.49                100
23831               6.62                100
7550                6.63                100

[3513 rows x 2 columns]

You should have a total of $3513$ rated anime

**Q. Which Anime (by ID) was rated most often? How often? What was it's average rating?**

Answer:

Now, we will import a second dataset called `anime.csv`

In this one, the first column is the `anime_id` (which is unique)

You should have $10228$ rows, each describing one anime

In [4]:
## Import the dataset; it looks like this

name  \
anime_id                                                      
32281                                        Kimi no Na wa.   
5114                       Fullmetal Alchemist: Brotherhood   
28977                                              Gintama°   
9253                                            Steins;Gate   
9969                                          Gintama&#039;   
...                                                     ...   
11095                               Zouressha ga Yatte Kita   
7808                       Zukkoke Knight: Don De La Mancha   
28543     Zukkoke Sannin-gumi no Hi Asobi Boushi Daisakusen   
18967             Zukkoke Sannin-gumi: Zukkoke Jikuu Bouken   
13455                                      Zumomo to Nupepe   

                                                      genre   type episodes  
anime_id                                                                     
32281                  Drama, Romance, School, Supernatural  Movie        1  
5114      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64  
28977     Action, Comedy, Historical, Parody, Samurai, S...     TV       51  
9253                                       Sci-Fi, Thriller     TV       24  
9969      Action, Comedy, Historical, Parody, Samurai, S...     TV       51  
...                                                     ...    ...      ...  
11095                                             Adventure  Movie        1  
7808                 Adventure, Comedy, Historical, Romance     TV       23  
28543                                           Drama, Kids    OVA        1  
18967                            Comedy, Historical, Sci-Fi    OVA        1  
13455                                                Comedy     TV       32  

[10228 rows x 4 columns]

We want to add the ratings from the `anime_ratings` dataframe onto the `anime` dataframe

Every rating in `anime_ratings` has a corresponding entry in `anime`

However, since we removed the entries of lesser rated anime (in the `anime_ratings` dataset), there are anime in the `anime` dataframe without corresponding entries in `anime_ratings`

**Q. If we want to keep all of the information about anime (regardless of whether we have the rating or not), what type of join should we us, considering `anime` as the left and `anime_ratings` as the right dataset?**

Answer:

**Q. If we want to only keep anime which have ratings, what type of join should we use?**

Answer:

Perform a join to keep only anime which have ratings

You should have 3,513 rows

In [5]:
## Here you need to explore the proper join to add the new info to your rating table
## In Python it would take one line of code
## In tableau, you would need to connect the two columns as done in Lab 1

name  \
anime_id                                     
32281                       Kimi no Na wa.   
5114      Fullmetal Alchemist: Brotherhood   
28977                             Gintama°   
9253                           Steins;Gate   
9969                         Gintama&#039;   
...                                    ...   
19315                                 Pupa   
28929                       Vampire Holmes   
16608                              Shitcom   
413                        Hametsu no Mars   
3287          Tenkuu Danzai Skelter+Heaven   

                                                      genre   type episodes  \
anime_id                                                                      
32281                  Drama, Romance, School, Supernatural  Movie        1   
5114      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
28977     Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
9253                                       Sci-Fi, Thriller     TV       24   
9969      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                     ...    ...      ...   
19315                        Fantasy, Horror, Psychological     TV       12   
28929                Comedy, Mystery, Supernatural, Vampire     TV       12   
16608                                       Comedy, Romance    ONA        1   
413                                          Horror, Sci-Fi    OVA        1   
3287                                          Mecha, Sci-Fi    OVA        1   

          Average Rating  Number of Ratings  
anime_id                                     
32281               9.43               1961  
5114                9.32              21494  
28977               9.45               1188  
9253                9.26              17151  
9969                9.27               3115  
...                  ...                ...  
19315               4.22               2135  
28929               4.19                133  
16608               3.21                192  
413                 2.47               1024  
3287                2.08                210  

[3513 rows x 6 columns]

Sort the data frame by `Average Rating` (descending)

In [6]:
## Sorting!

name  \
anime_id                                              
28977                                      Gintama°   
32281                                Kimi no Na wa.   
820                            Ginga Eiyuu Densetsu   
5114               Fullmetal Alchemist: Brotherhood   
9969                                  Gintama&#039;   
...                                             ...   
28929                                Vampire Holmes   
1657      Byston Well Monogatari: Garzey no Tsubasa   
16608                                       Shitcom   
413                                 Hametsu no Mars   
3287                   Tenkuu Danzai Skelter+Heaven   

                                                      genre   type episodes  \
anime_id                                                                      
28977     Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
32281                  Drama, Romance, School, Supernatural  Movie        1   
820                          Drama, Military, Sci-Fi, Space    OVA      110   
5114      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
9969      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                     ...    ...      ...   
28929                Comedy, Mystery, Supernatural, Vampire     TV       12   
1657                                                Fantasy    OVA        3   
16608                                       Comedy, Romance    ONA        1   
413                                          Horror, Sci-Fi    OVA        1   
3287                                          Mecha, Sci-Fi    OVA        1   

          Average Rating  Number of Ratings  
anime_id                                     
28977               9.45               1188  
32281               9.43               1961  
820                 9.39                803  
5114                9.32              21494  
9969                9.27               3115  
...                  ...                ...  
28929               4.19                133  
1657                4.01                110  
16608               3.21                192  
413                 2.47               1024  
3287                2.08                210  

[3513 rows x 6 columns]

**Q. Which anime was most highly rated?**

Answer:

Sort the data frame by `Number of Ratings` (descending)

In [7]:
## Again, sorting!

name  \
anime_id                                                      
1535                                             Death Note   
11757                                      Sword Art Online   
16498                                    Shingeki no Kyojin   
1575                        Code Geass: Hangyaku no Lelouch   
6547                                           Angel Beats!   
...                                                     ...   
1081      Space Pirate Captain Herlock: Outside Legend -...   
7550                           Pokemon: Pikachu Tanken Club   
1932                                         Yes! Precure 5   
2249                                              Amon Saga   
503                                          Eien no Aselia   

                                                      genre     type episodes  \
anime_id                                                                        
1535      Mystery, Police, Psychological, Supernatural, ...       TV       37   
11757             Action, Adventure, Fantasy, Game, Romance       TV       25   
16498          Action, Drama, Fantasy, Shounen, Super Power       TV       25   
1575      Action, Mecha, Military, School, Sci-Fi, Super...       TV       25   
6547            Action, Comedy, Drama, School, Supernatural       TV       13   
...                                                     ...      ...      ...   
1081        Action, Adventure, Drama, Sci-Fi, Seinen, Space      OVA       13   
7550                       Adventure, Comedy, Fantasy, Kids  Special        1   
1932                 Action, Fantasy, Magic, School, Shoujo       TV       49   
2249                             Action, Adventure, Fantasy      OVA        1   
503                               Adventure, Drama, Fantasy      OVA        2   

          Average Rating  Number of Ratings  
anime_id                                     
1535                8.83              34226  
11757               8.14              26310  
16498               8.73              25290  
1575                8.93              24126  
6547                8.55              23565  
...                  ...                ...  
1081                7.78                100  
7550                6.63                100  
1932                7.38                100  
2249                5.76                100  
503                 5.78                100  

[3513 rows x 6 columns]

**Q. Which anime had the most ratings?**

Answer:

The types of anime are `Movie`, `Music`, `ONA`, `OVA`, `Special` and `TV`

Use data aggregation to summarise by type

In [8]:
## In Python, it is better if you drop first the name and genre columns so that you can group by type
## Tools like R also raise warnings, but excel and tableau won't mind when grouping
## Then again, aggregate with the count and the median for number of ratings,
## Finally, rename columns and sort!

Median Number of Ratings per Anime  Number of Anime
type                                                        
TV                                    921.0             1713
OVA                                   349.5              602
Movie                                 719.0              555
Special                               350.0              513
ONA                                   439.0               93
Music                                 255.0               37